dynamic_programming/Unbounded_0_1_Knapsack.cpp

```diff
@@ -0,0 +1,153 @@
+ * @return The maximum value that can be obtained for the given index and capacity.
+ */
+int KnapSackFilling(std::uint16_t i, std::uint16_t W, 
+                    const std::vector<std::uint16_t>& val, 
+                    const std::vector<std::uint16_t>& wt, 
+                    std::vector<std::vector<int>>& dp) {
+    if (i == 0) {
+        if (wt[0] <= W) {
+            return (W / wt[0]) * val[0]; // Take as many of the first item as possible
+        } else {
+            return 0; // Can't take the first item
+        }
+    }
+    if (dp[i][W] != -1) return dp[i][W]; // Return result if available
+
+    int nottake = KnapSackFilling(i - 1, W, val, wt, dp); // Value without taking item i
+    int take = 0;
+    if (W >= wt[i]) {
+        take = val[i] + KnapSackFilling(i, W - wt[i], val, wt, dp); // Value taking item i
+    }
+    return dp[i][W] = std::max(take, nottake); // Store and return the maximum value
+}
+
+/**
+ * @brief Wrapper function to initiate the unbounded knapsack calculation.
+ *
+ * @param N Number of items.
+ * @param W Maximum weight capacity of the knapsack.
+ * @param val Vector of values corresponding to the items.
+ * @param wt Vector of weights corresponding to the items.
+ * @return The maximum value that can be obtained for the given capacity.
+ */
+int unboundedKnapsack(std::uint16_t N, std::uint16_t W, 
+                      const std::vector<std::uint16_t>& val, 
+                      const std::vector<std::uint16_t>& wt) {
+    if(N==0)return 0; // Expect 0 since no items
+    std::vector<std::vector<int>> dp(N, std::vector<int>(W + 1, -1)); // Initialize memoization table
+    return KnapSackFilling(N - 1, W, val, wt, dp); // Start the calculation
+}
+
+} // unbounded_knapsack
+
+} // dynamic_programming
+
+/**
+ * @brief Function to run test cases for the unbounded knapsack implementation.
+ * 
+ * @details This function tests the implementation with predefined test cases.
+ * It asserts the expected outcomes to ensure correctness of the algorithm.
+ */
+static void tests() {
+    // Test Case 1
+    std::uint16_t N1 = 4;  // Number of items
+    std::vector<std::uint16_t> wt1 = {1, 3, 4, 5}; // Weights of the items
+    std::vector<std::uint16_t> val1 = {6, 1, 7, 7}; // Values of the items
+    std::uint16_t W1 = 8; // Maximum capacity of the knapsack
+    // Test the function and assert the expected output
+    assert(unboundedKnapsack(N1, W1, val1, wt1) == 48);
+    std::cout << "Maximum Knapsack value " << unboundedKnapsack(N1, W1, val1, wt1) << std::endl;
+
+    // Test Case 2
+    std::uint16_t N2 = 3; // Number of items
+    std::vector<std::uint16_t> wt2 = {10, 20, 30}; // Weights of the items
+    std::vector<std::uint16_t> val2 = {60, 100, 120}; // Values of the items
+    std::uint16_t W2 = 5; // Maximum capacity of the knapsack
+    // Test the function and assert the expected output
+    assert(unboundedKnapsack(N2, W2, val2, wt2) == 0);
```

Comment:
add a test case where this returns negative values since return type is negative

Reply:
@realstealthninja According to the problem constraints, both the weights and values of items are strictly greater than zero, which ensures the result will always be non-negative. Therefore, adding a test case that returns a negative value is invalid for this problem. If you're seeing negative values in your output, it likely indicates an issue with the implementation or the problem definition itself.

Reply:
> @realstealthninja According to the problem constraints, both the weights and values of items are strictly greater than or equal to zero, which ensures the result will always be non-negative. Therefore, adding a test case that returns a negative value is invalid for this problem. If you're seeing negative values in your output, it likely indicates an issue with the implementation or the problem definition itself.

Therefore the return value of the function should not be an int. it should be an unsigned integer.

Reply:
Sure @realstealthninja, changes are done accordingly.